### PR TITLE
Threshold Command

### PR DESCRIPTION
**Pull request by billwpierce**
I Took the threshold command and added it to standard. Will run the command when it gets past a certain point. Once merged, we should add to 2017-Code.

### CODE REVIEW

File: commands/ThresholdCommand.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package org.usfirst.frc4904.standard.commands;
+
+
+import java.util.function.Supplier;
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: We should add a javadoc (or remove this empty one).

File: commands/ThresholdCommand.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public class ThresholdCommand extends Command {
+	protected final Command command;
+	protected final Supplier<Double> axis;
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: Now that this is a general command, this isn't necessarily an `axis` so we should refactor this to something else (maybe `supplier`).

File: commands/ThresholdCommand.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	protected final double threshold;
+	protected final boolean invert;
+
+	public ThresholdCommand(Command command, Supplier<Double> axis, double threshold, boolean invert) {
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: Is it worth it to make this work with other comparable types, not just `Double`? (i.e. `Integer`) I'm not sure.

File: commands/ThresholdCommand.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		this.threshold = threshold;
+		this.invert = invert;
+	}
+
+	public ThresholdCommand(Command command, Supplier<Double> axis, double threshold) {
+		this(command, axis, threshold, false);
+	}
+
+	protected boolean pastThreshold(double value) {
+		return (!invert && value >= threshold) || (invert && value <= threshold);
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: This is pedantic, but I think we can do this without doing the `value >= threshold` twice:
```
(value >= threshold) ^ invert
```
where the `^` is XOR (exclusive or, one or the other but not both).
We want to return true if the value is greater than the threshold or if we're inverted but not both.

File: commands/ThresholdCommand.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		this.axis = axis;
+		this.threshold = threshold;
+		this.invert = invert;
+	}
+
+	public ThresholdCommand(Command command, Supplier<Double> axis, double threshold) {
+		this(command, axis, threshold, false);
+	}
+
+	protected boolean pastThreshold(double value) {
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: Should this take in a value or should it directly call `axis.get()`?

File: commands/ThresholdCommand.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+		return (!invert && value >= threshold) || (invert && value <= threshold);
+	}
+
+	@Override
+	protected void execute() {
+		if (pastThreshold(axis.get()) && !command.isRunning()) {
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: @carturn did we experience weirdness with `isRunning` in the past? I seem to recall some issue with it not being triggered the first tick?

File: commands/ThresholdCommand.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			command.start();
+		} else if (!pastThreshold(axis.get()) && command.isRunning()) {
+			command.cancel();
+		}
+	}
+
+	@Override
+	protected boolean isFinished() {
+		return false;
```
Comment (c0d3rman) [inline commands/ThresholdCommand.java]: Perhaps we should add a version that, when the supplier reaches the threshold, starts the command and ends immediately? (And one that does the same thing with cancelling a command?)

**Comment by billwpierce** [comment]
@c0d3rman I fixed the problems.

File: commands/ThresholdCommand.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+		this.threshold = threshold;
+		this.invert = invert;
+	}
+
+	public ThresholdCommand(Command command, Supplier<T> axis, T threshold) {
+		this(command, axis, threshold, false);
+	}
+
+	protected boolean pastThreshold() {
+		return ((supplier.get().compareTo(threshold) >= 0) ^ invert);
```
Comment (carturn) [inline commands/ThresholdCommand.java]: Do not use bitwise operators for logical operations. I recommend !=.

**Pull request merged**